src/basic/unit-file.c

```diff
@@ -273,6 +274,34 @@ int unit_file_build_name_map(
                         return log_oom();
         }
 
+        /* Go over all our search paths, chase their symlinks and store the
+         * result in the expanded_search_paths list.
+         *
+         * This is important for cases where any of the unit directories itself
+         * are symlinks into other directories and would therefore cause all of
+         * the unit files to be recognized as linked units.
+         *
+         * This is important for distributions such as NixOS where most paths
+         * in /etc/ are symlinks to some other location on the filesystem (e.g.
+         * into /nix/store/).
+         */
+        STRV_FOREACH(dir, (char**) lp->search_path) {
+                _cleanup_free_ char *resolved_dir = NULL;
+                r = strv_extend(&expanded_search_paths, *dir);
+                if (r < 0)
+                        return log_oom();
+
+                r = chase_symlinks(*dir, NULL, 0, &resolved_dir, NULL);
+                if (r < 0) {
+                        if (r != -ENOENT)
+                                log_warning_errno(r, "Failed to resolve symlink %s, ignoring: %m", *dir);
+                        continue;
+                }
+
+                if (strv_consume(&expanded_search_paths, TAKE_PTR(resolved_dir)) < 0)
+                        return log_oom();
+        }
+
         STRV_FOREACH(dir, (char**) lp->search_path) {
```

Comment:
hmm, so why not just iterate through the reduced search path here?

i.e it appears to me the "expanded" search path should be called the "reduced" search path, and also maybe we should call strv_uniq() or so, to maybe strip duplicates. And then after all that we shoud operate purely on that and basically forget about original search path while building the map.

Reply:
>hmm, so why not just iterate through the reduced search path here?

I don't really know anymore. IIRC I was just proposing the minimal solution here to get some feedback on the approach. You might be correct that we can change all the `search_path` usages to the new list of paths. My only concern would be that in cases where e.g. `/etc/systemd/system` is already a symlink to another location we would then never see that path in `systemctl {status,cat} <unit>`? Arguably that would be the correct information but perhaps a bit misleading to the unaware?

Reply:
I think we need both. I'm rebasing the patch now and I added an comment that describes this.